editor/py/serve.py

```diff
@@ -38,8 +38,37 @@
 # pylint: disable=invalid-name,no-member,inconsistent-return-statements
 app = flask.Flask(__name__, template_folder='../html')
 
-# Root directory for the editor database (persistent storage).
-DB_ROOT = os.environ.get('ORD_EDITOR_DB', 'db')
+# Root directory for the editor filesystem.
+_DB_ROOT = os.getenv('ORD_EDITOR_DB', 'db')
+# Failure message for check_path.
+_CHECK_PATH_MESSAGE = ''
+_CHECK_PATH_CODE = 403
+
+
+def check_path(path, root=None, user=None):
+    """Checks that a path is safe.
+
+    Specifically, we are looking for directory traversal attacks.
+
+    Args:
+        path: Path to check.
+        root: Root path of the editor filesystem. Defaults to _DB_ROOT.
+        user: String user name. Defaults to flask.g.user.
+
+    Returns:
+        The normalized absolutized path.
+
+    Raises:
+        ValueError: If the path is not deemed safe.
+    """
+    if root is None:
+        root = _DB_ROOT
+    if user is None:
+        user = flask.g.user
+    abspath = os.path.abspath(path)
+    if not abspath.startswith(os.path.abspath(os.path.join(root, user))):
```

Comment:
Is `flask.g.user` necessarily safe? I feel like `os.path.join(root, user)` isn't the safest to check against. Since we have no concept of permissions, I would probably suggest that you just check if the path starts with `_DB_ROOT`

Reply:
That's fair; will do.

Reply:
Changed to only check for `_DB_ROOT`. I left the checks in the user section that don't rely on user input just for consistency.

Reply:
LGTM